less-11

Taras48 opened the pull request:
добавил комментарий и круд для него, переписал на JPA

kataus commented on src/main/kotlin/com/library/booklibrary/model/Book.kt:
```diff
@@ -1,21 +1,47 @@
+    )
+    @Fetch(FetchMode.SUBSELECT)
     var authors: MutableList<Author> = mutableListOf(),
-    var gener: Genre? = null
```
А жанры зачем пропали?

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/model/Book.kt:
```diff
@@ -1,21 +1,47 @@
 )
 
-fun Book.bookToBookDto()=
+fun Book.bookToBookDto() =
```
Нарушение single responsibility. В итоге растет связность, API не отделить от модели

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/model/Book.kt:
```diff
@@ -1,21 +1,47 @@
+        this.id,
+        this.name,
+        this.authors.map { it.authorToAuthorDto() }.toMutableList(),
+        this.comments.map { it.commentToCommentDto() }.toMutableList()
```
Опасно, у нас много чего ленивого

Taras48 replied:
не понял коммент, вроде по дефолту есть листы, в чем опасность?

kataus replied:
По дефолту да. Но если загрузить ленивую зависимость - тут будет прокси, а не реальный список. И прилетит LazyInitException

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/service/BookServiceImpl.kt:
```diff
@@ -5,23 +5,26 @@ import com.library.booklibrary.dto.BookDto
 
-    override fun getAllBooks()=
+    override fun getAllBooks() =
         bookDao.getAllBooks()?.map { it.bookToBookDto() }
```
Круто, транзакция мало того что тут явно не имеет пользы. Так ещё с правом на коммит, больно дорого получится

kataus replied:
Так зачем тут транзакция? Вопрос открытый. На обновление транзакций нет, например delete не закоммитит никаких изменений в базу. А на чтение мы будем поднимать транзакцию с правом на коммит, что весьма дорого

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/dao/AuthorDaoImpl.kt:
```diff
@@ -0,0 +1,44 @@
+            .resultList
+
+    override fun deleteAuthorById(id: Long) =
+        em.createQuery("delete from Author where id = :id")
```
А чем em.remove плох? Такой подход достаточно опасен, потому что работает в обход кэша и сущность может там остаться, а из базы уже удалили

Taras48 replied:
а при remove на нужно сбрасывать контекст чтобы в базе удаление произошло, в данном случае если мы заперсистим книги после удаления авторов, то удаленные авторы будут опять в базе при remove

kataus replied:
Зачем сбрасываать контект? транзакция закрылась, из базы удалилось. При ручном удалении обозначенная проблема - реальна. Если довериться  ORM - звучит неправдоподобно. При remove уже ORM почистит и не допустит

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/dao/AuthorDaoImpl.kt:
```diff
@@ -0,0 +1,44 @@
+            .executeUpdate()
+
+    override fun updateAuthorNameById(id: Long, name: String) =
+        em.createQuery(
```
ORM сделет эту работу для нас

Taras48 replied:
не понял комментарий, пожалуйста поясните что исправить

kataus replied:
зачем вручную в dao обновлять?

Taras48 replied:
не понял, а как тогда нужно?

kataus replied:
Через бизнес-логику. Ручные обновления не нужны, это задача  ORM, на занятии показывали обновление сущностей

kataus commented on src/main/kotlin/com/library/booklibrary/dao/BookDaoImpl.kt:
```diff
@@ -1,128 +1,44 @@
+import javax.persistence.EntityManager
+import javax.persistence.PersistenceContext
 
 @Repository
```
Что тут используется от аннотации @repository? Она несколько больше чем @Component или @Service. Если она появляется -то мы ожидаем использование её особенностей

Taras48 replied:
обеспечение доп семантики, да, конечно сдесь не используется эта аннотация по полной, но вы же писали что относится как к боевому проекту, возможно при дальнейшей разработке нужно будет работать с исключениями

Taras48 replied:
ну и опять же общепринятые принципы, по моему мнению если поставлю @component или @service, возможно введу в заблуждения других разработчиков

kataus replied:
Чем можно ввести в заблуждение? Аннотации редко кто смотрит, снаружи видно только имя класса/интерфейса, оно полностью описывает задачу.
Когда-то в будущем может и понадобится (что ещё маловероятно), а вот нагрузку на свою работу будет требовать уже сейчас

kataus commented on src/main/kotlin/com/library/booklibrary/service/CommentServiceImpl.kt:
```diff
@@ -0,0 +1,33 @@
+    override fun saveComment(comment: CommentDto) =
+        commentDao.saveComment(comment.commentDtoToComment())
+            .commentToCommentDto()
+}
```
А давайте добавим ещё метод "получить комментарии по книге" При этом не надо добавлять методы для этого в dao, в модели уже есть всё необходимое

Taras48 replied:
добавил

kataus commented on src/main/kotlin/com/library/booklibrary/shellCommand/AuthorCommands.kt:
```diff
@@ -0,0 +1,47 @@
+    private val outputConsoleService: OutputConsoleService,
+) {
+
+//    @ShellMethod(key = ["db"])
```
мертвый код не храним, пока он нас под собой не похоронил

kataus commented on src/main/kotlin/com/library/booklibrary/model/Book.kt:
```diff
@@ -1,21 +1,43 @@
-this.name,
-this.authors.map { it.authorToAuthorDto() }.toMutableList(),
-this.gener?.genreToGenreDto(),
+    @ManyToOne(targetEntity = Genre::class, fetch = FetchType.LAZY, optional = false)
```
Тут замена на LAZY спорна, не зря по умолчанию EAGER

Taras48 replied:
поправил

kataus commented on src/main/kotlin/com/library/booklibrary/service/BookServiceImpl.kt:
```diff
@@ -2,26 +2,27 @@ package com.library.booklibrary.service
 
-    override fun insertBook(book: BookDto) =
-        bookDao.insertBook(book.bookDtoToBook())
+    override fun saveBook(book: BookDto) =
```
В базе ничего не останется по итогам работы данного метода

kataus commented on src/main/kotlin/com/library/booklibrary/service/CommentServiceImpl.kt:
```diff
@@ -0,0 +1,35 @@
+        commentDao.getAllComments()
+            ?.map { it.commentToCommentDto() }
+
+    override fun getCommentsByBookId(id: Long) =
```
LazyInitException вылетит

Taras48 replied:
поправил

Commit:
less-11
внес правки

Commit:
less-11
поправил связь

Commit:
less-11
добавил транзакции

kataus commented on src/main/kotlin/com/library/booklibrary/dao/BookDaoImpl.kt:
```diff
@@ -1,128 +1,49 @@
-
-    private fun getAuthorsForBooks(): MutableMap<Long, MutableList<Author>>? {
-        val map = jdbc.query(
+    override fun updateBookNameById(id: Long, name: String) =
```
ORM - это не маппинг. Это другая концепция. Мы работаем с объектом ничего не зная о его хранении. В идеале мы можем даже поменять тип базы без серьезных доработок. Если нет ручных привязок к конкретной базе

Taras48 replied:
поправил

Commit:
less-11
перенес update в сервис

kataus commented on src/main/kotlin/com/library/booklibrary/service/BookServiceImpl.kt:
```diff
@@ -2,26 +2,36 @@ package com.library.booklibrary.service
+    override fun updateBookNameById(id: Long, name: String){
+        bookDao.findBookById(id)?.let {
+            it.name = name
+            bookDao.saveBook(it)
```
👍🏻

Commit:
less-11
поправил методы сервиса

kataus commented on src/main/kotlin/com/library/booklibrary/dao/BookDaoImpl.kt:
```diff
@@ -1,128 +1,58 @@
-                on b.id = g.book_id
-                """.trimMargin(),
-            EXTRACTOR
+        em.createQuery(
```
а чем  em.find плох? то же самое сделает

kataus commented on src/main/kotlin/com/library/booklibrary/service/CommentServiceImpl.kt:
```diff
@@ -0,0 +1,40 @@
+            ?.map { it.commentToCommentDto() }
+
+    override fun getCommentsByBookId(id: Long) =
+        bookService.findBookById(id)?.comments
```
Обратитесь тут в dao за книгой, чтобы не мучаться с dto. Тут от find надо только ссылку на прокси с комментариями, который потом разрешить прям тут

Commit:
less-11
поправил поиск книги и коммента по книги

Commit:
less-11
поправил поиск книги и коммента по книги

Commit:
less-11
добавил комменты обратно в книгу, добавил транзакции на полкчение комментов по книге

Commit:
less-11
добавил комменты обратно в книгу, добавил транзакции на полкчение комментов по книге

Commit:
Merge branch 'master' of https://github.com/Taras48/Book-library into less-11

 Conflicts:
	src/main/kotlin/com/library/booklibrary/dao/AuthorDao.kt
	src/main/kotlin/com/library/booklibrary/dao/AuthorDaoImpl.kt
	src/main/kotlin/com/library/booklibrary/dao/BookDao.kt
	src/main/kotlin/com/library/booklibrary/dao/BookDaoImpl.kt
	src/main/kotlin/com/library/booklibrary/dao/CommentDao.kt
	src/main/kotlin/com/library/booklibrary/dao/CommentDaoImpl.kt
	src/main/kotlin/com/library/booklibrary/dao/GenreDao.kt
	src/main/kotlin/com/library/booklibrary/dao/GenreDaoImpl.kt
	src/main/kotlin/com/library/booklibrary/model/Book.kt
	src/main/kotlin/com/library/booklibrary/model/Genre.kt
	src/main/kotlin/com/library/booklibrary/service/AuthorService.kt
	src/main/kotlin/com/library/booklibrary/service/AuthorServiceImpl.kt
	src/main/kotlin/com/library/booklibrary/service/BookServiceImpl.kt
	src/main/kotlin/com/library/booklibrary/service/CommentService.kt
	src/main/kotlin/com/library/booklibrary/service/CommentServiceImpl.kt
	src/main/kotlin/com/library/booklibrary/service/GenreService.kt
	src/main/kotlin/com/library/booklibrary/service/GenreServiceImpl.kt
	src/test/kotlin/com/library/booklibrary/BookDaoImplTest/BookDaoImplTest.kt

Commit:
less-11
